internal/bundle/utils.go

```diff
@@ -88,26 +167,61 @@ func LoadWasmResolversFromStore(ctx context.Context, store storage.Store, txn st
 }
 
 // LoadBundleFromDisk loads a previously persisted activated bundle from disk
-func LoadBundleFromDisk(path, name string, bvc *bundle.VerificationConfig) (*bundle.Bundle, error) {
-	bundlePath := filepath.Join(path, name, "bundle.tar.gz")
+func LoadBundleFromDisk(path string, opts *LoadOptions) (*bundle.Bundle, error) {
+	// if a bundle package exists, use that as it might contain the bundle etag which
+	// is not stored in the legacy bundle file. This can help avoid unnecessary bundle
+	// downloads.
+	bundlePackagePath := filepath.Join(path, PackageFileName)
+	if _, err := os.Stat(bundlePackagePath); err == nil {
+		f, err := os.Open(filepath.Join(bundlePackagePath))
+		if err != nil {
+			return nil, fmt.Errorf("failed to open bundle package file: %w", err)
+		}
 
+		var bundlePackage bundlePackage
+		err = json.NewDecoder(f).Decode(&bundlePackage)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode bundle package file: %w", err)
+		}
+
+		r := bundle.NewReader(bytes.NewReader(bundlePackage.Bundle))
```

Comment:
Can you please help me understand how this works? In `SaveBundleToDisk` we're setting the `RawBundle` field and here we're calling `bytes.NewReader` on the `Bundle` field. So is the point of the `Bundle` field essentially to trigger a read on the `bundlePackage` struct where `RawBundle` is processed?

Reply:
The `Bundle` field on `bundlePackage` needs to be a `[]byte` so that we can unmarshal the data correctly from the persisted JSON field which is base64 encoded.

`RawBundle` is only used at write time, and supports efficient writing of bundle data from the original `io.Reader`.

So the overall process is:
* Create a `bundlePackage` with the ETag and `RawBundle` set. `RawBundle` is the `io.Reader` from the downloader.
* `io.Copy` the `bundlePackage` to disk. This uses `func (bp *bundlePackage) Read...` to manually format out a JSON representation of `bundlePackage` while efficiently copying from the input reader to disk without any extra allocations. This uses a `io.MultiReader` to manually build the JSON. `Bundle` is not used here and is only for unmarshalling.
* After this, the bundlePackage exists on disk as a JSON file.
* When loading, `LoadBundleFromDiskWithOptions` will not use the `RawBundle` field at all, and only use the `bundlePackage` options (ETag) to configure the bundle reader before reading the `Bundle`  data (set during the unmarshalling) into a `bundle.Bundle` struct.